### PR TITLE
aaronmassey/fuel gauge emul backend api

### PR DESCRIPTION
**Pull request by aaronemassey**
Add a minimal fuel gauge emulator backend API to make scenario testing easier.

### CODE REVIEW

File: include/zephyr/drivers/emul_fuel_gauge.h [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+#include <zephyr/drivers/emul.h>
+#include <zephyr/drivers/fuel_gauge.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief Fuel gauge backend emulator APIs
+ * @defgroup fuel_gauge_emulator_backend fuel gauge backed emulator APIs
```
Comment (keith-zephyr) [inline include/zephyr/drivers/emul_fuel_gauge.h:25]: Add this doxygen group to the fuel_gauge.rst doc.

Comment (aaronemassey) [reply include/zephyr/drivers/emul_fuel_gauge.h:25]: Ahh! Thanks!

File: drivers/fuel_gauge/sbs_gauge/emul_sbs_gauge.c [hunk 2]
```diff
@@ -220,10 +233,49 @@ static int sbs_gauge_emul_transfer_i2c(const struct emul *target, struct i2c_msg
 	return rc;
 }
 
+static int emul_sbs_fuel_gauge_set_battery_charging(const struct emul *target, uint32_t mV, int mA)
+{
+	struct sbs_gauge_emul_data *data = target->data;
+
+	if (mV == 0 || mA == 0)
+		return -EINVAL;
```
Comment (yperess) [inline drivers/fuel_gauge/sbs_gauge/emul_sbs_gauge.c:241]: Nit: code style

File: drivers/fuel_gauge/sbs_gauge/emul_sbs_gauge.c [hunk 3]
```diff
@@ -220,10 +233,49 @@ static int sbs_gauge_emul_transfer_i2c(const struct emul *target, struct i2c_msg
...
+
+	data->batt_state.mA = mA;
+	data->batt_state.mV = mV;
```
Comment (keith-zephyr) [inline drivers/fuel_gauge/sbs_gauge/emul_sbs_gauge.c:244]: One thought as a follow up effort.  
Add a backend hook to set the charging voltage and current.  The emulator could scheduling a recurring work item and slowly adjust the "real" voltage and current as appropriate.

**Review by keith-zephyr — DISMISSED** [review]

File: drivers/fuel_gauge/sbs_gauge/emul_sbs_gauge.c (hunk 3 above)
Comment (aaronemassey) [reply drivers/fuel_gauge/sbs_gauge/emul_sbs_gauge.c:244]: Good idea! I'll make a github issue so it's tracked.

**Review by yperess — DISMISSED** [review]
Minor nit only, can merge and fix next week imo

File: tests/drivers/fuel_gauge/sbs_gauge/src/test_sbs_gauge.c [hunk 4]
```diff
@@ -344,5 +348,29 @@ ZTEST_USER_F(sbs_gauge_new_api, test_get_buffer_props__returns_ok)
...
+	struct fuel_gauge_get_property props[] = {
+		{
+			.property_type = FUEL_GAUGE_VOLTAGE,
+		},
+		{
+			.property_type = FUEL_GAUGE_CURRENT,
+		},
+	};
+
+	zassume_ok(emul_fuel_gauge_set_battery_charging(fixture->sbs_fuel_gauge, 5000, 3000));
```
Comment (perowle) [inline tests/drivers/fuel_gauge/sbs_gauge/src/test_sbs_gauge.c:364]: Might consider making 5000 and 3000 symbols, since they're referenced more than once

**Review by perowle — DISMISSED** [review]

**Comment by aaronemassey** [comment]
@yperess @keith-zephyr @perowle 

Surprisingly the build_all sensor tests don't enable ztest. So for now I can't embed the reset rule in the emulator. Are we okay with just having the backend emulator API merged without the reset?

**Comment by keith-zephyr** [comment]
> @yperess @keith-zephyr @perowle
> 
> Surprisingly the build_all sensor tests don't enable ztest. So for now I can't embed the reset rule in the emulator. Are we okay with just having the backend emulator API merged without the reset?

Will the lack of a reset cause any test flakiness?

**Comment by aaronemassey** [comment]
> > @yperess @keith-zephyr @perowle
> > Surprisingly the build_all sensor tests don't enable ztest. So for now I can't embed the reset rule in the emulator. Are we okay with just having the backend emulator API merged without the reset?
> 
> Will the lack of a reset cause any test flakiness?

Not the the test as it stands now. I could add a before/after in the test now if you'd like.

**Comment by yperess** [comment]
@aaronemassey I'm actually about to ask @tristan-google to convert that test to a generic sensor test. You can either:

1. Add ztest to the test
2. Wrap the reset with `ifdef CONFIG_ZTEST`

**Comment by aaronemassey** [comment]
Btw, just want to put out that it is no big deal at all if this does *not* get merged this week :)

File: include/zephyr/drivers/emul_fuel_gauge.h [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+/**
+ * @brief Set charging for fuel gauge assoicated battery.
+ *
+ * Set how much the battery associated with a fuel gauge IC is charging or discharging. Where
+ * voltage is always positive and a positive or negative current denotes charging or discharging,
+ * respectively.
+ *
+ * @param target Pointer to the emulator structure for the fuel gauge emulator instance.
+ * @param mV Millivolts describing the battery voltage.
+ * @param mA Milliamps describing the battery current where negative is discharging.
```
Comment (keith-zephyr) [inline include/zephyr/drivers/emul_fuel_gauge.h:51]: Why is the emulator using mV/mA while the fuel gauge API uses uA/uV?

**Comment by aaronemassey** [comment]
> @aaronemassey I'm actually about to ask @tristan-google to convert that test to a generic sensor test. You can either:
> 
> 1. Add ztest to the test
> 2. Wrap the reset with `ifdef CONFIG_ZTEST`

Done. I also have to stub out ZTEST_DMEM since I need it for userspace tests.

**Review by yperess — APPROVED** [review]
minor nit

**Review by keith-zephyr — APPROVED** [review]

**Review by teburd — APPROVED** [review]

**Pull request merged**